src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs

```diff
@@ -2803,6 +2805,52 @@ private static bool IsReallyAType(TypeSymbol type)
                 !type.IsErrorType() &&
                 (type.SpecialType != SpecialType.System_Void);
         }
+
+        private void AddOrMerge(Dictionary<TypeSymbol, TypeSymbol> dict, TypeSymbol entry)
+        {
+            if (dict.TryGetValue(entry, out TypeSymbol found))
+            {
+                MergeAndReplace(dict, found, entry);
+            }
+            else
+            {
+                dict.Add(entry, entry);
+            }
+        }
+
+        private void MergeAndReplace(Dictionary<TypeSymbol, TypeSymbol> dict, TypeSymbol old, TypeSymbol @new)
+        {
+            // We make an exception when @new is dynamic, for backwards compatibility 
+            if (@new.IsDynamic())
+            {
+                return;
+            }
+
+            TypeSymbol merged = MergeTupleNames(MergeDynamic(old, @new, _conversions.CorLibrary), @new);
+            dict.Remove(old);
```

Comment:
>dict.Remove(old); [](start = 12, length = 17)

Is it possible that we get in a situation when the value is no longer in the dictionary, should we be adding it back in this case?